be/src/exec/parquet_builder.cpp

```diff
@@ -0,0 +1,351 @@
+}
+
+void ParquetBuildHelper::build_compression_type(parquet::WriterProperties::Builder& builder,
+                                                const TCompressionType::type& compression_type) {
+    switch (compression_type) {
+    case TCompressionType::SNAPPY: {
+        builder.compression(parquet::Compression::SNAPPY);
+        break;
+    }
+    case TCompressionType::GZIP: {
+        builder.compression(parquet::Compression::GZIP);
+        break;
+    }
+    case TCompressionType::BROTLI: {
+        builder.compression(parquet::Compression::BROTLI);
+        break;
+    }
+    case TCompressionType::ZSTD: {
+        builder.compression(parquet::Compression::ZSTD);
+        break;
+    }
+    case TCompressionType::LZ4: {
+        builder.compression(parquet::Compression::LZ4);
+        break;
+    }
+    case TCompressionType::LZO: {
+        builder.compression(parquet::Compression::LZO);
+        break;
+    }
+    case TCompressionType::BZIP2: {
+        builder.compression(parquet::Compression::BZ2);
+        break;
+    }
+    case TCompressionType::DEFAULT_COMPRESSION: {
+        builder.compression(parquet::Compression::UNCOMPRESSED);
+        break;
+    }
+    default:
+        builder.compression(parquet::Compression::UNCOMPRESSED);
+    }
+}
+
+void ParquetBuilder::_generate_rg_writer() {
+    if (_rg_writer == nullptr) {
+        _rg_writer = _file_writer->AppendBufferedRowGroup();
+        _cur_written_rows = 0;
+    }
+}
+
+#define DISPATCH_PARQUET_NUMERIC_WRITER(WRITER, COLUMN_TYPE, NATIVE_TYPE)                                         \
+    ParquetBuilder::_generate_rg_writer();                                                                        \
+    parquet::WRITER* col_writer = static_cast<parquet::WRITER*>(_rg_writer->column(i));                           \
+    col_writer->WriteBatch(                                                                                       \
+            num_rows, nullable ? def_level.data() : nullptr, nullptr,                                             \
+            reinterpret_cast<const NATIVE_TYPE*>(down_cast<const COLUMN_TYPE*>(data_column)->get_data().data())); \
+    _buffered_values_estimate[i] = col_writer->EstimatedBufferedValueBytes();
+
+Status ParquetBuilder::add_chunk(Chunk* chunk) {
+    if (!chunk->has_rows()) {
+        return Status::OK();
+    }
+
+    size_t num_rows = chunk->num_rows();
+    for (size_t i = 0; i < chunk->num_columns(); i++) {
+        auto& col = chunk->get_column_by_index(i);
+        bool nullable = col->is_nullable();
+        auto null_column = nullable && down_cast<NullableColumn*>(col.get())->has_null()
+                                   ? down_cast<NullableColumn*>(col.get())->null_column()
+                                   : nullptr;
+        const auto data_column = ColumnHelper::get_data_column(col.get());
+
+        std::vector<int16_t> def_level(num_rows, 1);
```

Comment:
better to extract a helper function for calculating def_level and rep_level, because the algorithm is a little bit complicated if we support nested types.

Reply:
we will implement it in detail later